File: algorithm/python/dp/lcs.py
#-*-encoding:utf-8-*-
from tracer import *
st1 = 'AGGTAB'
st2 = 'GXTXAYB'
m = len(st1)
n = len(st2)
f = []
for i in range(m+1):
    f.append([])
    for j in range(n+1):
        f[i].append(0)

tracer1 = Array1DTracer('string1').setData(st1)
tracer2 = Array1DTracer('string2').setData(st2)
tracer3 = Array2DTracer('MemoTable').setData(f)
logger = LogTracer('Output')


def dp():
    # Build the memo table in bottom up fashion
    for i in range(1, m+1):
        for j in range(1, n+1):
            if st1[i-1] == st2[j-1]:
                tracer1.select(i-1).wait()
                tracer2.select(j-1).wait()
                tracer3.select(i-1,j-1).wait()

                f[i][j] = f[i-1][j-1] + 1

                tracer1.deselect(i-1)
                tracer2.deselect(j-1)
                tracer3.deselect(i-1, j-1)
            else:
                tracer3.select(i-1, j).wait()
                tracer3.select(i, j-1).wait()

                f[i][j] = max(f[i-1][j], f[i][j-1])

                tracer3.deselect(i-1, j)
                tracer3.deselect(i, j-1)
        tracer3.notify(i, j, f[i][j]).wait().denotify(i, j)


def get_ans():
    ans = ''
    i = m
    j = n
    while i >= 1 and j >= 1:
        tracer3.select(i, j).wait()
        if st1[i-1] == st2[j-1]:
            tracer1.select(i-1).wait()
            tracer2.select(j-1).wait()

            ans = st1[i-1] + ans
        elif f[i-1][j] > f[i][j-1]:
            i -= 1
        else:
            j -= 1
    ans1 = u'Longest Common Subsequence Length is %d' % f[m][n]
    ans2 = u'Longest Common Subsequence is %s' % ans
    logger.Print(ans1)
    logger.Print(ans2)


if __name__ == '__main__':
    dp()
    get_ans()
    tracerlist.info()

File: algorithm/javascript/dp/lcs.js
var string1 = 'AGGTAB';
var string2 = 'GXTXAYB';
var m = string1.length;
var n = string2.length;
var A = new Array (m+1);
for (var i = 0; i < m+1; i++ ) {
	A[i] = new Array (n+1);
}

var tracer1 = new Array1DTracer ( 'String 1')._setData ( string1 );
var tracer2 = new Array1DTracer ( 'String 2')._setData ( string2 );
var tracer3 = new Array2DTracer ( 'Memo Table')._setData ( A );
var logger = new LogTracer ();


var i,j;

// Build the memo table in bottom up fashion 
 for( i = 0; i <= m; i++ ) {
 	for( j = 0; j <= n; j++ ) {
 		if ( i === 0 || j === 0 ) {
 			A[i][j] = 0;
 		} else if ( string1[i-1] == string2[j-1] ) {
 			tracer1._select ( i-1 )._wait ();
 			tracer2._select ( j-1 )._wait ();
 			tracer3._select ( i-1, j-1 )._wait ();
 			
 			A[i][j] = A[i-1][j-1] + 1;
 			
 			tracer1._deselect ( i-1 );
 			tracer2._deselect ( j-1 );
 			tracer3._deselect ( i-1, j-1 );
 		} else {
 			tracer3._select ( i-1, j )._wait ();
 			tracer3._select ( i, j-1 )._wait ();
 			
 			if( A[i-1][j] > A[i][j-1] ) {
 				A[i][j] = A[i-1][j];
 			} else {
 				A[i][j] = A[i][j-1];
 			}
 			
 			tracer3._deselect ( i-1, j );
 			tracer3._deselect ( i, j-1 );
 		}
 		tracer3._notify ( i, j , A[i][j] )._wait ();
 		tracer3._denotify( i, j );
 	}
 }

var finalString = '';
i=m;
j=n;
while( i>=1 && j>=1 ) {

	tracer3._select ( i, j )._wait ();
	if( string1[i-1] == string2[j-1] ) {
		tracer1._select ( i-1 )._wait ();
 		tracer2._select ( j-1 )._wait ();
 		
		finalString = string1[i-1] + finalString;
		i--;
		j--;
	} else if( A[i-1][j] > A[i][j-1] ) {
		i--;
	} else {
		j--;
	}
}

logger._print ( 'Longest Common Subsequence Length is ' + A[m][n] );
logger._print ( 'Longest Common Subsequence is ' + finalString );